Major refactoring to match future MRtrix 3.1.0

Lestropie opened the pull request:
- Divide the code across multiple source files.
- Move source files to new location according to MRtrix 3.1.0 filesystem arrangement.
- Build against development branch of MRtrix3 (future version 3.1.0). This includes incorporation of multiple upstream features, such as metadata handling fixes and dwicat dealing with subtle differences in voxel grids.
- Make use of dwidenoise2.
- Perform gradient non-linearity distortion correction if the requisite deformation field is made available.

Closes #136.
Closes #134 (though currently not "genuinely" compiled as an external project; the code is injected into an MRtrix3 installation prior to `cmake` configuration).
Closes #133.
Closes #132.
Closes #131.
Closes #130.
Closes #123.
Closes #122.
Closes #114.
Closes #106 (exposed as command-line option).
Closes #64.
Closes #45.

- [x] Remove manual implementation of steps underlying `dwibiasnormmask`; assume presence of `dwibiasnormmask` given script no longer seeks to be executable outside of the container environment and is intrinsically tied to MRtrix 3.1.
- [x] Expose cubic first-level `eddy` model (#129) as command-line option; appropriate triggering of this functionality may depend on DICOM metadata that does not make it to BIDS
- [ ] Test `preproc` level analysis on a broader spectrum of input datasets
- [ ] Test `participant` and `group` level analyses
- [x] Utilise complex data for Gibbs ringing removal given that capability is available in the imported MRtrix3 version.

Lestropie committed:
Major refactoring to match future MRtrix 3.1.0

- Divide the code across multiple source files.
- Move source files to new location according to MRtrix 3.1.0 filesystem arrangement.
- Build against development branch of MRtrix3 (future version 3.1.0). This includes incorporation of multiple upstream features, such as metadata handling fixes and dwicat dealing with subtle differences in voxel grids.
- Make use of dwidenoise2.
- Perform gradient non-linearity distortion correction if the requisite deformation field is made available.

Lestropie committed:
Remove branch with manual implementation of dwibiasnormmask contents

Lestropie committed:
Add preprepco-level options -eddy_cubicflm, -eddy_mbs

aaroncapon commented on mrtrix3_connectome/execute.py:
```diff
@@ -76,15 +76,17 @@ def execute(): #pylint: disable=unused-variable
                         session_to_process,
                         preproc_shared,
                         app.ARGS.t1w_preproc,
-                        app.ARGS.concat_denoise,
```
Tried running mrtrix3_connectome 'preproc' using this branch but am getting a TypeError due to a missing parameter when it goes to run preproc.run.run_preproc(). The run_preproc() is still expecting the concat_denoise argument but that has been removed here. Looks to still be used in the function (line 363).

Lestropie replied:
Should just be an incomplete refactoring: 208426bb3f3895778d4ebe7b68b5833f2d351ca4

Lestropie committed:
Preproc: Fix refactoring of "concat_denoise"

Lestropie committed:
Fix erroneous usage of eddy MBS functionality

Erroneously set Shared class member variable name already in use, resulting in engagement on that functionality based in its availability rather than user request.

Lestropie committed:
Code cleanup #137

Lestropie committed:
Use SynthStrip for T1w brain extraction

Closes #138.

Lestropie committed:
Fix use of FSL for derivation of brain mask only

Where the user provides as input a derivatives directory that only provides a pre-processed T1-weighted image and not a corresponding mask, FSL's fsl_anat command may produce images for which the field of view in specifically the inferior portion of the image has been reduced. This would otherwise lead to downstream problems where the mask extracted from this directory and the input pre-processed T1-weighted image are expected to reside on the same voxel grid, initially by the mrhistmatch command. This fix resamples the FSL-derived brain mask back onto the full voxel grid of the input pre-processed T1-weighted image.

Lestropie committed:
Preproc: Fix import of complex DWI data

Lestropie committed:
Merge pull request #139 from bids-apps/t1w_synthstrip

Use SynthStrip for T1w brain extraction

Lestropie committed:
Preproc: Run Gibbs ringing removal on complex data

Commit:
Fix T1w import with -t1w_preproc but no mask

Lestropie committed:
Update dwidenoise2

Requisite update that prevents the manifestation of NaNs in the output denoised series, which would result in complete image corruption by mrdegibbs.

Lestropie commented:
@aaroncapon Updated (couldn't find any record of updating the `dwidenoise2` tip within this tool)

Lestropie committed:
Permit image provided via -t1w_preproc to not have "_desc-preproc"

Lestropie committed:
Preproc: Fix calculation of MPorder

Previous code would result in implausible MPorder value (and immediate eddy failure) if the multiband acceleration factor was at least 4 as evidenced by the contents of metadata field "SliceTiming", but the metadata did not contain field "MultibandAccelerationFactor", as the requested order would exceed the number of slice groups. This change explicitly computes the latter if it is absent from the input metadata.

Lestropie committed:
Preproc: Fixes to inter-modal registration

- Use eroded masks for histogram matching and registration to mitigate hypointensities at the peripheries of the masks leading to hyperintensities in the contrast-inverted images.
- Perform contrast inversion using a linear rather than reciprocal transform, taking into account the prospect of values that are negative / zero / small in magnitude (which could be introduced by eg. gradient non-linearity distortion correction of the T1w) that have a deleterious effect on the reciprocal transform.
Closes #142.
Closes #141.

Lestropie committed:
Fix variable name error

Error in modifications in 54cff609a1c8a11d283dbe0fb23c125328aaddc8.

Lestropie committed:
Merge pull request #143 from bids-apps/fix_registration_GDC

Fix interaction between GDC and registration

Lestropie committed:
Preproc: More robust GDC image parsing

Prevents erroneous collision detection between "MAGNETOM Cima.X" and "MAGNETOM Cima.X Fit".
